### ISSUE TITLE
NVIDIA development directly on Tensor cores - specifically the GB10 on the DGX Spark

### ISSUE THREAD

**Issue by obriensystems** [ObrienlabsDev/blog#153]
This one in collaboration with Gemini pro - for throughput planning
see https://github.com/ObrienlabsDev/performance/issues/47

https://gemini.google.com/app/ffe7490def8a744e
"
how do I directly program the neural processing units in apple silicon on for example the M4max or M3Ultra or the tensor cores in a GB10 on an nvidia dgx spark
"

<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>

<meta http-equiv="Content-Style-Type" content="text/css">
<title></title>
<meta name="Generator" content="Cocoa HTML Writer">
<meta name="CocoaVersion" content="2685.3">
<style type="text/css">
p.p1 {margin: 0.0px 0.0px 12.0px 0.0px; font: 12.0px Helvetica; -webkit-text-stroke: #000000}
p.p2 {margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Helvetica; color: #6d6d6d; -webkit-text-stroke: #6d6d6d; min-height: 14.0px}
p.p3 {margin: 0.0px 0.0px 14.0px 0.0px; font: 14.0px Helvetica; -webkit-text-stroke: #000000}
p.p4 {margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Helvetica; -webkit-text-stroke: #000000; min-height: 14.0px}
p.p5 {margin: 0.0px 0.0px 12.0px 0.0px; font: 12.0px Times; -webkit-text-stroke: #000000; min-height: 14.0px}
p.p6 {margin: 0.0px 0.0px 15.9px 0.0px; font: 12.0px Helvetica; -webkit-text-stroke: #000000}
p.p7 {margin: 0.0px 0.0px 0.0px 0.0px; font: 12.0px Helvetica; -webkit-text-stroke: #000000}
li.li1 {margin: 0.0px 0.0px 12.0px 0.0px; font: 12.0px Helvetica; -webkit-text-stroke: #000000}
span.s1 {font-kerning: none}
span.s2 {font: 16.0px Helvetica; font-kerning: none}
table.t1 {margin: 0.0px 0.0px 32.0px 0.0px}
td.td1 {width: 98.7px; margin: 0.5px 0.5px 0.5px 0.5px; border-style: solid; border-width: 1.0px 1.0px 1.0px 1.0px; border-color: #000000 #000000 #000000 #000000; padding: 1.0px 1.0px 1.0px 1.0px}
td.td2 {width: 168.7px; margin: 0.5px 0.5px 0.5px 0.5px; border-style: solid; border-width: 1.0px 1.0px 1.0px 1.0px; border-color: #000000 #000000 #000000 #000000; padding: 1.0px 1.0px 1.0px 1.0px}
td.td3 {width: 220.7px; margin: 0.5px 0.5px 0.5px 0.5px; border-style: solid; border-width: 1.0px 1.0px 1.0px 1.0px; border-color: #000000 #000000 #000000 #000000; padding: 1.0px 1.0px 1.0px 1.0px}
ul.ul1 {list-style-type: disc}
</style>
</head>
<body>
<p class="p1"><span class="s1">The short answer is that while you can program NVIDIA Tensor Cores <b>explicitly and down to the assembly level</b>, you strictly <b>cannot</b> program Apple's Neural Engine (ANE) directly in the same sense. Apple's NPU is a "black box" designed to be accessed only through high-level compiler layers.</span></p>
<p class="p1"><span class="s1">Here is the breakdown of how to get as close to the metal as possible on both architectures.</span></p>
<p class="p2"><span class="s1"></span><br></p>
<p class="p3"><span class="s1"><b>1. Apple Silicon (M3 Ultra / M4 Max)</b></span></p>
<p class="p1"><span class="s1"><b>Status:</b> <b>Closed / Undocumented ISA</b><b></b></span></p>
<p class="p1"><span class="s1">You cannot write C++ or Assembly code that runs directly on the NPU (Neural Engine).</span><span class="s2"><sup>1</sup></span><span class="s1"> Apple treats the ANE as an implementation detail of the OS. There is no public instruction set architecture (ISA) and no direct driver interface.</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p5"><span class="s1"></span><br></p>
<p class="p6"><span class="s1"><b>The "Official" Lowest Level: MPSGraph</b></span></p>
<p class="p1"><span class="s1">The closest you can get to "metal" programming on Apple Silicon for AI is <b>Metal Performance Shaders Graph (MPSGraph)</b>. This bypasses CoreML's high-level abstractions and lets you build the compute graph manually.</span><span class="s2"><sup>2</sup><sup></sup></span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p5"><span class="s1"></span><br></p>
<ul class="ul1">
<li class="li1"><span class="s1"><b>How it works:</b> You define a graph of mathematical operations (matrix multiplications, convolutions).</span><span class="s2"><sup>3</sup></span><span class="s1"> The OS compiler then decides <i>at runtime</i> whether to run this on the GPU, the NPU, or the AMX (Apple Matrix Coprocessor on the CPU).</span><span class="s2"><sup>4</sup></span><span class="s1"><br>
<br>
<br>
</span></li>
<li class="li1"><span class="s1"><b>Constraint:</b> You cannot <i>force</i> execution on the NPU. You can only "hint" it. The NPU mostly accelerates very specific quantized operations (Int8/FP16 convolutions).</span></li>
</ul>
<p class="p7"><span class="s1">Swift</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">// Example: Creating a graph in Swift (closest you get to "direct")</span></p>
<p class="p7"><span class="s1">import MetalPerformanceShadersGraph</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">let graph = MPSGraph()</span></p>
<p class="p7"><span class="s1">let a = graph.placeholder(shape: [1024, 1024], dataType: .float32, name: "A")</span></p>
<p class="p7"><span class="s1">let b = graph.placeholder(shape: [1024, 1024], dataType: .float32, name: "B")</span></p>
<p class="p7"><span class="s1">let c = graph.matrixMultiplication(primary: a, secondary: b, name: "C")</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">// The OS driver compiles this graph and dispatches it to GPU or NPU</span></p>
<p class="p6"><span class="s1"><b>The "Hacker" Level: Reverse Engineering</b></span></p>
<p class="p1"><span class="s1">There have been community efforts (such as the geohot/tinygrad project or anemll) to reverse-engineer the ANE ISA.</span></p>
<ul class="ul1">
<li class="li1"><span class="s1">They found the ANE is effectively a massive, statically scheduled spatial dataflow machine.</span></li>
<li class="li1"><span class="s1"><b>Current State:</b> It is extremely brittle. Apple changes the ANE ISA with every chip generation (M1 $\to$ M2 $\to$ M3), breaking these tools immediately. It is not viable for production.</span></li>
</ul>
<p class="p2"><span class="s1"></span><br></p>
<p class="p3"><span class="s1"><b>2. NVIDIA (Blackwell GB200 / "GB10")</b></span></p>
<p class="p1"><span class="s1"><b>Status:</b> <b>Open / Fully Programmable</b><b></b></span></p>
<p class="p1"><span class="s1"><i>Note: You likely mean the </i><b><i>GB200</i></b><i> (Grace Blackwell) architecture. "DGX Spark" is possibly a reference to </i><b><i>DGX SuperPOD</i></b><i> or a specific cluster configuration.</i><i></i></span></p>
<p class="p1"><span class="s1">On NVIDIA hardware, you have full access to the Tensor Cores.</span><span class="s2"><sup>5</sup></span><span class="s1"> You can bypass high-level frameworks (PyTorch/TensorFlow) and program them using CUDA C++ or inline PTX assembly.</span><span class="s2"><sup>6</sup><sup></sup></span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p5"><span class="s1"></span><br></p>
<p class="p6"><span class="s1"><b>Level 1: CUDA WMMA API (Warp Matrix Multiply Accumulate)</b></span></p>
<p class="p1"><span class="s1">This is the standard "direct" way to program Tensor Cores without writing assembly. You use the nvcuda::wmma namespace to load data into special fragments and trigger the hardware matrix multiply.</span><span class="s2"><sup>7</sup><sup></sup></span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p5"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">C++</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">#include &lt;mma.h&gt;</span></p>
<p class="p7"><span class="s1">using namespace nvcuda;</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">__global__ void wmma_kernel(half *a, half *b, float *c) {</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>// Declare fragments (registers specifically for Tensor Cores)</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::fragment&lt;wmma::matrix_a, 16, 16, 16, half, wmma::row_major&gt; a_frag;</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::fragment&lt;wmma::matrix_b, 16, 16, 16, half, wmma::col_major&gt; b_frag;</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::fragment&lt;wmma::accumulator, 16, 16, 16, float&gt; c_frag;</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>// Load data from memory to Tensor Core registers</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::load_matrix_sync(a_frag, a, 16);</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::load_matrix_sync(b_frag, b, 16);</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>// Execute the hardware instruction</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::mma_sync(c_frag, a_frag, b_frag, c_frag);</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>// Store result</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>wmma::store_matrix_sync(c, c_frag, 16, wmma::mem_row_major);</span></p>
<p class="p7"><span class="s1">}</span></p>
<p class="p6"><span class="s1"><b>Level 2: Inline PTX (Assembly)</b></span></p>
<p class="p1"><span class="s1">If you are on the Blackwell architecture (Compute Capability 10.0+), you can use the newest mma instructions or the new <b>TMA (Tensor Memory Accelerator)</b> to asynchronously copy data directly from global memory into the Tensor Core shared memory clusters.</span><span class="s2"><sup>8</sup><sup></sup></span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p5"><span class="s1"></span><br></p>
<p class="p1"><span class="s1">You can inject assembly directly into your C++ code to use instructions that might not yet be exposed in the CUDA C++ API:</span></p>
<p class="p7"><span class="s1">C++</span></p>
<p class="p4"><span class="s1"></span><br></p>
<p class="p7"><span class="s1">asm volatile(</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>"mma.sync.aligned.m16n8k16.row.col.f32.f16.f16.f32 "</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>"{%0, %1, %2, %3}, {%4, %5, %6, %7}, {%8, %9}, {%10, %11, %12, %13};"</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>: "+f"(d0), "+f"(d1), "+f"(d2), "+f"(d3)</span></p>
<p class="p7"><span class="s1"><span class="Apple-converted-space">   </span>: "r"(a0), "r"(a1), "r"(a2), "r"(a3), "r"(b0), "r"(b1), "f"(c0), "f"(c1), "f"(c2), "f"(c3)</span></p>
<p class="p7"><span class="s1">);</span></p>
<p class="p6"><span class="s1"><b>Level 3: CUTLASS &amp; CuTe</b></span></p>
<p class="p1"><span class="s1">Because programming Tensor Cores directly involves complex shared-memory swizzling to avoid bank conflicts, NVIDIA engineers themselves use a template library called <b>CUTLASS</b> (and its core layout library <b>CuTe</b>).</span></p>
<ul class="ul1">
<li class="li1"><span class="s1">This is "direct programming" but with C++ templates handling the register allocation math.</span></li>
<li class="li1"><span class="s1">On Blackwell (GB200), this is critical because you need to coordinate the <b>TMA</b> (async copy engines) with the <b>WGMMA</b> (Warpgroup Matrix Multiply) instructions.</span></li>
</ul>
<p class="p3"><span class="s1"><b>Summary of Differences</b></span></p>

Feature | Apple Neural Engine (M3/M4) | NVIDIA Tensor Core (GB200)
-- | -- | --
Programmability | Indirect Only (Compiler driven) | Direct (C++ or Assembly)
Lowest Level API | MPSGraph / Core ML | CUDA WMMA / PTX Assembly
Instruction Set | Private (Undocumented) | Public (PTX / SASS)
Data Types | Mostly Int8 / FP16 | Int4, Int8, FP8, BF16, FP16, TF32, FP64
Memory Access | Managed by OS Cache | Explicit (User controls Shared Mem/TMA)

<p class="p1"><span class="s1"><b>Recommendation:</b><b></b></span></p>
<ul class="ul1">
<li class="li1"><span class="s1"><b>For Apple:</b> Use <b>Core ML</b> or <b>MPSGraph</b>. Do not attempt to program the ANE directly unless you are researching security/reverse-engineering; it is a moving target that will break your code in the next macOS update.</span></li>
<li class="li1"><span class="s1"><b>For NVIDIA:</b> If you want max performance, use <b>CUTLASS</b>. It generates the highly optimized pipeline (loading data -&gt; Tensor Core math -&gt; storing data) that is tedious to write manually in raw CUDA.</span></li>
</ul>
</body>
</html>
